### ISSUE TITLE
Deprecation warnings in lightning.pytorch.cli with jsonargparse 4.49+

### ISSUE THREAD

**Issue by adamjstewart** [Lightning-AI/pytorch-lightning#21900]
### Bug description

`lightning/pytorch/cli.py` uses a few features from jsonargparse that have recently been deprecated.

### What version are you seeing the problem on?

v2.6

### How to reproduce the bug

```python

```

### Error messages and logs

```
  /Users/Adam/torchgeo/.venv/lib/python3.14/site-packages/lightning/pytorch/cli.py:596: 
      ``ArgumentParser.add_instantiator`` was deprecated in v4.49.0 and will be removed in v5.0.0. Use the global
      function ``jsonargparse.add_instantiator`` instead.

  /Users/Adam/torchgeo/.venv/lib/python3.14/site-packages/lightning/pytorch/cli.py:612: 
      ``instantiate_classes`` was deprecated in v4.49.0 and will be removed in v5.0.0. Instead use ``instantiate``.

  /Users/Adam/torchgeo/.venv/lib/python3.14/site-packages/lightning/pytorch/cli.py:590: 
      skip_none parameter was deprecated in v4.49.0 and will be removed in v5.0.0. Instead use skip_unset.
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
	- GPU:               None
	- available:         False
	- version:           None
* Lightning:
	- lightning:         2.6.5
	- lightning-utilities: 0.15.3
	- pytorch-lightning: 2.6.5
	- segmentation_models_pytorch: 0.5.0
	- torch:             2.13.0
	- torchgeo:          0.10.0
	- torchmetrics:      1.9.0
	- torchvision:       0.28.0
* Packages:
	- Jinja2:            3.1.6
	- MarkupSafe:        3.0.3
	- PyYAML:            6.0.3
	- Pygments:          2.20.0
	- Sphinx:            9.1.0
	- accessible-pygments: 0.0.5
	- aenum:             3.1.17
	- affine:            3.0.0
	- aiohappyeyeballs:  2.7.1
	- aiohttp:           3.14.3
	- aiosignal:         1.4.0
	- alabaster:         1.0.0
	- annotated-types:   0.8.0
	- antlr4-python3-runtime: 4.9.3
	- anyio:             4.14.2
	- appnope:           0.1.4
	- asttokens:         3.0.2
	- attrs:             26.1.0
	- azure-core:        1.41.0
	- azure-storage-blob: 12.30.0
	- babel:             2.18.0
	- beautifulsoup4:    4.15.0
	- bleach:            6.4.0
	- braceexpand:       0.1.7
	- certifi:           2026.7.22
	- cffi:              2.1.1
	- cftime:            1.6.5
	- charset-normalizer: 3.5.0
	- click:             8.4.2
	- comm:              0.2.3
	- contourpy:         1.3.3
	- coverage:          7.15.4
	- cryptography:      50.0.0
	- cycler:            0.12.1
	- debugpy:           1.8.21
	- defusedxml:        0.7.1
	- docstring_parser:  0.18.0
	- docutils:          0.22.4
	- einops:            0.8.2
	- executing:         2.2.1
	- fastjsonschema:    2.22.1
	- filelock:          3.32.3
	- fonttools:         4.63.0
	- frozenlist:        1.8.0
	- fsspec:            2026.7.0
	- geopandas:         1.1.4
	- h11:               0.16.0
	- hf-xet:            1.6.0
	- httpcore:          1.0.9
	- httpx:             0.28.1
	- huggingface_hub:   1.27.0
	- hydra-core:        1.3.5
	- idna:              3.18
	- imagesize:         2.0.0
	- iniconfig:         2.3.0
	- ipykernel:         7.3.0
	- ipython:           9.16.1
	- ipython_pygments_lexers: 1.1.1
	- ipywidgets:        8.1.8
	- isodate:           0.7.2
	- jedi:              0.20.0
	- jsonargparse:      4.50.0
	- jsonnet:           0.22.0
	- jsonschema:        4.26.0
	- jsonschema-specifications: 2025.9.1
	- jupyter_client:    8.9.1
	- jupyter_core:      5.9.1
	- jupyterlab_pygments: 0.3.0
	- jupyterlab_widgets: 3.0.16
	- kiwisolver:        1.5.0
	- kornia:            0.8.3
	- kornia_rs:         0.1.14
	- laspy:             2.7.0
	- lightly:           1.5.26
	- lightly-utils:     0.0.2
	- lightning:         2.6.5
	- lightning-utilities: 0.15.3
	- markdown-it-py:    4.2.0
	- matplotlib:        3.11.1
	- matplotlib-inline: 0.2.2
	- mdit-py-plugins:   0.6.1
	- mdurl:             0.1.2
	- microsoft-aurora:  2.0.1
	- mistune:           3.3.4
	- mpmath:            1.3.0
	- multidict:         6.7.1
	- myst-parser:       5.1.0
	- nbclient:          0.11.0
	- nbconvert:         7.17.1
	- nbformat:          5.11.0
	- nbmake:            1.5.5
	- nbsphinx:          0.9.8
	- nest-asyncio2:     1.7.2
	- netCDF4:           1.7.4
	- networkx:          3.6.1
	- numpy:             2.5.2
	- olmoearth-pretrain-minimal: 0.0.6
	- omegaconf:         2.3.1
	- packaging:         26.3
	- pandas:            3.0.5
	- pandas-stubs:      3.0.5.260730
	- pandocfilters:     1.5.1
	- parso:             0.8.7
	- pathlib_abc:       0.5.2
	- pexpect:           4.9.0
	- pillow:            12.3.0
	- platformdirs:      4.11.3
	- pluggy:            1.6.0
	- prompt_toolkit:    3.0.53
	- propcache:         0.5.2
	- psutil:            7.2.2
	- ptyprocess:        0.7.0
	- pure_eval:         0.2.3
	- pyarrow:           25.0.1
	- pycocotools:       2.0.11
	- pycparser:         3.0
	- pydantic:          2.13.4
	- pydantic_core:     2.46.4
	- pydata-sphinx-theme: 0.20.0
	- pyogrio:           0.13.0
	- pyparsing:         3.3.2
	- pyproj:            3.7.2
	- pytest:            9.1.1
	- pytest-cov:        7.1.0
	- pytest-socket:     0.8.0
	- python-dateutil:   2.9.0.post0
	- pytorch-lightning: 2.6.5
	- pyzmq:             27.1.0
	- rasterio:          1.5.1
	- referencing:       0.37.0
	- requests:          2.34.2
	- rioxarray:         0.23.0
	- roman-numerals:    4.1.0
	- rpds-py:           2026.6.3
	- ruff:              0.16.3
	- safetensors:       0.8.0
	- scipy:             1.18.0
	- segmentation_models_pytorch: 0.5.0
	- setuptools:        84.0.0
	- shapely:           2.1.2
	- six:               1.17.0
	- snowballstemmer:   3.1.1
	- soupsieve:         2.9.2
	- sphinx-github-changelog: 2.3.0
	- sphinxcontrib-applehelp: 2.0.0
	- sphinxcontrib-devhelp: 2.0.0
	- sphinxcontrib-htmlhelp: 2.1.0
	- sphinxcontrib-jsmath: 1.0.1
	- sphinxcontrib-qthelp: 2.0.0
	- sphinxcontrib-serializinghtml: 2.0.0
	- stack-data:        0.6.3
	- sympy:             1.14.0
	- tenacity:          9.1.4
	- timm:              1.0.28
	- tinycss2:          1.5.1
	- torch:             2.13.0
	- torchgeo:          0.10.0
	- torchmetrics:      1.9.0
	- torchvision:       0.28.0
	- tornado:           6.5.8
	- tqdm:              4.70.0
	- traitlets:         5.16.1
	- ty:                0.0.71
	- types-geopandas:   1.1.4.20260807
	- types-rasterio:    1.5.0.20260810
	- types-requests:    2.33.0.20260712
	- types-shapely:     2.1.0.20260728
	- typeshed_client:   2.13.0
	- typing-inspection: 0.4.4
	- typing_extensions: 4.16.0
	- universal_pathlib: 0.3.10
	- urllib3:           2.7.0
	- wcwidth:           0.8.2
	- webdataset:        1.0.2
	- webencodings:      0.5.1
	- widgetsnbextension: 4.0.15
	- xarray:            2026.7.0
	- yarl:              1.24.5
* System:
	- OS:                Darwin
	- architecture:
		- 64bit
		- Mach-O
	- processor:         arm
	- python:            3.14.4
	- release:           25.5.0
	- version:           Darwin Kernel Version 25.5.0: Tue Jun  9 22:28:24 PDT 2026; root:xnu-12377.121.10~1/RELEASE_ARM64_T6020

</details>

cc @ethanwharris

**Comment by VenishPaneliya** [comment]
Hi, I'd like to work on this.

Plan: update the three deprecated jsonargparse calls in `src/lightning/pytorch/cli.py` (`ArgumentParser.add_instantiator`, `instantiate_classes`, and the `skip_none` parameter) to their 4.49+ replacements, while keeping compatibility with the minimum jsonargparse version Lightning currently supports, and add a test asserting no `DeprecationWarning` is raised from `cli.py`.

I'll confirm the supported version range before deciding between a straight swap and a version-gated shim. Happy to adjust the scope if you'd prefer a different direction.

**Comment by VenishPaneliya** [comment]
Looking into this, the replacements can't be adopted under the current version pin.

`requirements/pytorch/extra.txt` pins `jsonargparse[signatures,jsonnet] >=4.39.0, <4.48.0`, and `cli.py:39` enforces a `>=4.39` floor via `RequirementCache`. Per the jsonargparse changelog, `instantiate` and the global `add_instantiator` arrived in **v4.49.0**, and `skip_unset` in **v4.50.0** — all above the current ceiling. So swapping the calls directly would raise `AttributeError`/`TypeError` on every currently supported version. The warnings in the report come from running jsonargparse 4.50.0, which is outside the supported range.

There are also more call sites than the three listed:

- `skip_none`: lines 291, 307, 590
- `parser.add_instantiator`: lines 596, 601
- `parser.instantiate_classes`: lines 612, 904

(Worth noting `LightningCLI.instantiate_classes()` at line 610 is Lightning's own public method, unrelated to jsonargparse's deprecated one despite the shared name — lines 415-418 call it, and it shouldn't be renamed.)

Which direction would you prefer?

1. Raise the ceiling to allow >=4.49 and version-gate the calls so 4.39-4.48 keeps working
2. Raise the floor to >=4.49 and migrate cleanly, dropping the older range
3. Bump the ceiling only and defer the migration

Happy to implement whichever you pick — option 1 is the most conservative but adds compatibility branches in `cli.py`, option 2 is the cleanest diff if dropping 4.39-4.48 is acceptable for the next release.
